packages/dapi/lib/grpcServer/handlers/blockheaders-stream/getHistoricalBlockHeadersIteratorFactory.js

```diff
@@ -34,17 +36,45 @@ function getHistoricalBlockHeadersIteratorFactory(coreRpcApi) {
 
     for (let batchIndex = 0; batchIndex < numberOfBatches; batchIndex++) {
       const currentHeight = fromBlockHeight + batchIndex * MAX_HEADERS_PER_REQUEST;
-      const blocksToScan = getBlocksToScan(batchIndex, numberOfBatches, count);
 
-      const blockHash = await coreRpcApi.getBlockHash(currentHeight);
+      let blocksToScan = getBlocksToScan(batchIndex, numberOfBatches, count);
+      let blockHash = await coreRpcApi.getBlockHash(currentHeight);
+
+      const blockHeights = [...Array(blocksToScan).keys()]
+        .map((e, i) => currentHeight + i)
+
+      const cachedBlockHeaders = blockHeights.map((height) => cache.get(height))
+      const [firstCachedItem] = cachedBlockHeaders
+
+      let lastCachedIndex = -1
+
+      if (firstCachedItem) {
+        const firstMissingIndex = cachedBlockHeaders.indexOf(undefined)
+
+        // return cache if we do not miss anything
+        if (cachedBlockHeaders.filter(e => !!e).length === blocksToScan) {
```

Comment:
Not sure if thats a good way to check that

Reply:
Could you just check `firstMissingIndex === -1` instead?